// html-validate/dist/cjs/test-utils.js
'use strict';

var fs = require('fs');

function _interopDefault (e) { return e && e.__esModule ? e : { default: e }; }

var fs__default = /*#__PURE__*/_interopDefault(fs);

function transformFile(fn, filename, chain) {
  const data = fs__default.default.readFileSync(filename, "utf-8");
  const source = {
    filename,
    line: 1,
    column: 1,
    offset: 0,
    data
  };
  return transformSource(fn, source, chain);
}
function transformString(fn, data, chain) {
  const source = {
    filename: "inline",
    line: 1,
    column: 1,
    offset: 0,
    data
  };
  return transformSource(fn, source, chain);
}
function transformSource(fn, source, chain) {
  const defaultChain = (source2) => [source2];
  const context = {
    hasChain: (
      /* istanbul ignore next */
      () => true
    ),
    chain: chain ?? defaultChain
  };
  return Array.from(fn.call(context, source));
}

exports.transformFile = transformFile;
exports.transformSource = transformSource;
exports.transformString = transformString;
